amazon/ion/core.py

```diff
@@ -389,13 +464,48 @@ def __new__(cls, *args, **kwargs):
             if fractional_precision is not None and 1 > fractional_precision > MICROSECOND_PRECISION:
                 raise ValueError('Cannot construct a Timestamp with fractional precision of %d digits, '
                                  'which is out of the supported range of [1, %d].'
-                                 % (fractional_precision, MICROSECOND_PRECISION,))
+                                 % (fractional_precision, MICROSECOND_PRECISION))
             # Make sure we mask this before we construct the datetime.
             del kwargs[TIMESTAMP_FRACTION_PRECISION_FIELD]
+        if TIMESTAMP_FRACTIONAL_SECONDS_FIELD in kwargs:
+            fractional_seconds = kwargs.get(TIMESTAMP_FRACTIONAL_SECONDS_FIELD)
+            if fractional_seconds is not None:
+                fractional_seconds = Decimal(fractional_seconds)
+                if 0 > fractional_seconds >= 1:
+                    raise ValueError('Cannot construct a Timestamp with fractional seconds of %s, '
+                                     'which is out of the supported range of [0, 1).'
+                                     % str(fractional_seconds))
+            # Make sure we mask this before we construct the datetime.
+            del kwargs[TIMESTAMP_FRACTIONAL_SECONDS_FIELD]
+
+        if fractional_seconds is not None and (fractional_precision is not None or datetime_microseconds is not None):
+            raise ValueError('fractional_seconds cannot be specified '
+                             'when fractional_precision or microseconds are not None.')
+
+        if fractional_precision is not None and datetime_microseconds is None:
+            raise ValueError('datetime_microseconds cannot be None while fractional_precision is not None.')
+
+        if fractional_seconds is not None:
+            fractional_precision = min(fractional_seconds.as_tuple().exponent * -1, MICROSECOND_PRECISION)
+            datetime_microseconds = _fractional_seconds_to_microseconds(fractional_seconds)
+            lst = list(args)
+            lst[microsecond_argument_index] = datetime_microseconds
+            args = tuple(lst)
+        elif not (datetime_microseconds == 0 and fractional_precision is None):
+            if datetime_microseconds is None:
+                if len(args) > 6:
+                    datetime_microseconds = 0
+                    lst = list(args)
+                    lst[microsecond_argument_index] = datetime_microseconds
+                    args = tuple(lst)
+            else:
+                fractional_seconds = str(_scale_to_precision(datetime_microseconds, MICROSECOND_PRECISION))
+                fractional_seconds = Decimal(fractional_seconds[0:(fractional_precision + 2)])
```

Comment:
It's to make sure we have `fractional_seconds` at the right precision when we compute it from `datetime_microseconds` and `fractional_precision`. 
Example:
Let's say in the timestamp we want microseconds of 0.001000 with precision 3, so 0.001. This value will be stored in `datetime_microseconds` as 1000 and `fractional_precision` will be 3.
We want to compute the corresponding `fractional_seconds`. In line 502, what we get back from `_scale_to_precision` is 0.001000 and we stringify it.
Then in line 503 we truncate this string to `fractional_precision + 2` (to account for the `0.` in the beginning of the string), cast it to Decimal, then we end up with the correct `fractional_seconds` Decimal value of 0.001.

Reply:
I see. It may be more succinct and efficient (avoiding the round trip via `str`) to `_scale_to_precision(datetime_microseconds, MICROSECOND_PRECISION).quantize(precision_limit, decimal.ROUND_FLOOR)`, where `precision_limit` is `Decimal('0.' + '0' * (fractional_precision - 1) + '1')`. We can even have a static lookup table for `precision_limit` since `fractional_precision` is guaranteed to be [1, 6] at this point. (See: [Decimal.quantize](https://docs.python.org/2/library/decimal.html#decimal.Decimal.quantize))

Reply:
Done. Thank you.